src/Databases/DatabaseHDFS.cpp

```diff
@@ -0,0 +1,228 @@
+{
+    // Check if the table exists in the loaded tables map
+    {
+        std::lock_guard lock(mutex);
+        auto it = loaded_tables.find(name);
+        if (it != loaded_tables.end())
+            return it->second;
+    }
+
+    auto url = getTablePath(name);
+
+    checkUrl(url, context_, true);
+
+    // call TableFunctionHDFS
+    auto args = makeASTFunction("hdfs", std::make_shared<ASTLiteral>(url));
+
+    auto table_function = TableFunctionFactory::instance().get(args, context_);
+    if (!table_function)
+        return nullptr;
+
+    // TableFunctionHDFS throws exceptions, if table cannot be created
+    auto table_storage = table_function->execute(args, context_, name);
+    if (table_storage)
+        addTable(name, table_storage);
+
+    return table_storage;
+}
+
+StoragePtr DatabaseHDFS::getTable(const String & name, ContextPtr context_) const
+{
+    // rethrow all exceptions from TableFunctionHDFS to show correct error to user
+    if (auto storage = getTableImpl(name, context_))
+        return storage;
+    throw Exception(ErrorCodes::UNKNOWN_TABLE, "Table {}.{} doesn't exist", backQuoteIfNeed(getDatabaseName()), backQuoteIfNeed(name));
+}
+
+StoragePtr DatabaseHDFS::tryGetTable(const String & name, ContextPtr context_) const
+{
+    try
+    {
+        return getTableImpl(name, context_);
+    }
+    catch (const Exception & e)
+    {
+        // Ignore exceptions thrown by TableFunctionHDFS, which indicate that there is no table
+        if (e.code() == ErrorCodes::BAD_ARGUMENTS)
+            return nullptr;
+        if (e.code() == ErrorCodes::ACCESS_DENIED)
+            return nullptr;
+        if (e.code() == ErrorCodes::DATABASE_ACCESS_DENIED)
+            return nullptr;
+        if (e.code() == ErrorCodes::FILE_DOESNT_EXIST)
+            return nullptr;
+        if (e.code() == ErrorCodes::UNACCEPTABLE_URL)
+            return nullptr;
+        if (e.code() == ErrorCodes::HDFS_ERROR)
+            return nullptr;
+        if (e.code() == ErrorCodes::CANNOT_EXTRACT_TABLE_STRUCTURE)
+            return nullptr;
+        throw;
+    }
+    catch (const Poco::URISyntaxException &)
+    {
+        return nullptr;
+    }
+}
+
+ASTPtr DatabaseHDFS::getCreateDatabaseQuery() const
+{
+    auto settings = getContext()->getSettingsRef();
+    ParserCreateQuery parser;
+
+    const String query = fmt::format("CREATE DATABASE {} ENGINE = HDFS('{}')", backQuoteIfNeed(getDatabaseName()), source);
+    ASTPtr ast = parseQuery(parser, query.data(), query.data() + query.size(), "", 0, settings.max_parser_depth);
+
+    if (const auto database_comment = getDatabaseComment(); !database_comment.empty())
+    {
+        auto & ast_create_query = ast->as<ASTCreateQuery &>();
+        ast_create_query.set(ast_create_query.comment, std::make_shared<ASTLiteral>(database_comment));
+    }
+
+    return ast;
+}
+
+void DatabaseHDFS::shutdown()
+{
+    Tables tables_snapshot;
+    {
+        std::lock_guard lock(mutex);
```

Comment:
In [DatabaseCommon](https://github.com/ClickHouse/ClickHouse/blob/0e2156c77693ba0cfdd3c62b113f7f3a57ca2ffa/src/Databases/DatabasesCommon.cpp#L284-L285), it is not recommended to do shutdown of the table under lock. Snapshot allows to avoid race condition

I don't think this applies to temporary tables, but still...